jwst/ami/oifits.py

```diff
@@ -0,0 +1,763 @@
+            B1 = np.sqrt(u1coord[k] ** 2 + v1coord[k] ** 2)
+            B2 = np.sqrt(u2coord[k] ** 2 + v2coord[k] ** 2)
+            B3 = np.sqrt(u3coord[k] ** 2 + v3coord[k] ** 2)
+            bl_cp.append(np.max([B1, B2, B3]))  # rad-1
+        bl_cp = np.array(bl_cp)
+
+        flagVis = [False] * self.nbl
+        flagT3 = [False] * self.ncp
+
+        # do the things done by populate_nrm here
+        # average or don't, and get uncertainties
+        # Unwrap phases
+        shift2pi = np.zeros(self.cp.shape)
+        shift2pi[self.cp >= 6] = 2 * np.pi
+        shift2pi[self.cp <= -6] = -2 * np.pi
+        self.cp -= shift2pi
+
+        if self.method not in ['mean','median','multi']:
+            self.method = 'median'
+        # set these as attributes (some may exist and be overwritten)
+        if self.method == 'multi':
+            self.vis2 = self.fa2.T
+            self.e_vis2 = np.zeros(self.vis2.shape)
+            self.visamp = self.fa.T
+            self.e_visamp = np.zeros(self.visamp.shape)
+            self.visphi = self.fp.T
+            self.e_visphi = np.zeros(self.visphi.shape)
+            self.cp = self.cp.T
+            self.e_cp = np.zeros(self.cp.shape)
+            self.camp = self.ca.T
+            self.e_camp = np.zeros(self.camp.shape)
+            self.pist = self.pistons.T
+            self.e_pist = np.zeros(self.pist.shape)
+
+        # apply sigma-clipping to uncertainties
+        # sigma_clipped_stats returns mean, median, stddev. nsigma=3, niters=5
+        elif self.method == 'median':
+            _, self.vis2, self.e_vis2 = sigma_clipped_stats(self.fa2, axis=0)
+            _, self.visamp, self.e_visamp = sigma_clipped_stats(self.fa, axis=0)
+            _, self.visphi, self.e_visphi = sigma_clipped_stats(self.fp, axis=0)
+            _, self.cp, self.e_cp = sigma_clipped_stats(self.cp, axis=0)
+            _, self.camp, self.e_camp = sigma_clipped_stats(self.ca, axis=0)
+            _, self.pist, self.e_pist = sigma_clipped_stats(self.pistons, axis=0)
+
+        else: # take the mean
+            self.vis2, _, self.e_vis2 = sigma_clipped_stats(self.fa2, axis=0)
+            self.visamp, _, self.e_visamp = sigma_clipped_stats(self.fa, axis=0)
+            self.visphi, _, self.e_visphi = sigma_clipped_stats(self.fp, axis=0)
+            self.cp, _, self.e_cp = sigma_clipped_stats(self.cp, axis=0)
+            self.camp, _, self.e_camp = sigma_clipped_stats(self.ca, axis=0)
+            self.pist, _, self.e_pist = sigma_clipped_stats(self.pistons, axis=0)
+
+        # prepare arrays for OI_ARRAY ext
+        self.staxy = info4oif.ctrs_inst
+        N_ap = len(self.staxy)
+        tel_name = ['A%i' % x for x in np.arange(N_ap)+1]
+        sta_name = tel_name
+        diameter = [0] * N_ap
+
+        staxyz = []
+        for x in self.staxy:
+            a = list(x)
+            line = [a[0], a[1], 0]
+            staxyz.append(line)
+
+        sta_index = np.arange(N_ap) + 1
```

Comment:
The current truth files are failing oifits validation (using the oifits-check from [oifitslib](https://github.com/jsy1001/oifitslib)).

Here are the errors for `jw01093012001_03102_00001_nis_a3002_ami-oi.fits`:
```
*** Does not conform to the OIFITS standard:
Reference to missing array element, 27 occurrences:-
    OI_VIS #1 record 1
    OI_VIS #1 record 2
    OI_VIS #1 record 3
    OI_VIS #1 record 4
    OI_VIS #1 record 5
    OI_VIS #1 record 6
    OI_VIS2 #1 record 1
    OI_VIS2 #1 record 2
    OI_VIS2 #1 record 3
    [List truncated]
*** Does not conform to the OIFITS standard:
OI_T3 table may contain unnormalised triple product amplitude, 24 occurrences:-
    OI_T3 #1 record 3 channel 1
    OI_T3 #1 record 5 channel 1
    OI_T3 #1 record 6 channel 1
    OI_T3 #1 record 7 channel 1
    OI_T3 #1 record 8 channel 1
    OI_T3 #1 record 9 channel 1
    OI_T3 #1 record 10 channel 1
    OI_T3 #1 record 11 channel 1
    OI_T3 #1 record 13 channel 1
    [List truncated]
```

The errors might be related to this line where `sta_index` is computed. The `OI_ARRAY` table in the file does not contain a line for `sta_index` 0. However the first 6 rows of `OI_VIS` and `OI_VIS2` list `sta_index` 0 as does many rows in `OI_T3`.

Reply:
I missed a bit of reformatting; seems that the OIFITS standard wants these indices to be 1-indexed rather than 0-indexed. Just fixed that!